course1/course_en.md

```diff
@@ -0,0 +1,199 @@
+3. **Write the test cases**: What is the expected behavior for normal inputs? What can be the potential abnormal inputs?
+4. **Implement the program**: Usually, it is necessary to break down the problem into smaller subproblems and repeat the above process for each of them.
+
+---
+
+# An Example of Program Design
+
+> There are `num_bottles` water bottles that are initially full of water. You can exchange `num_exchange` empty water bottles from the market with one full water bottle.
+> The operation of drinking a full water bottle turns it into an empty bottle.
+> Given the two integers `num_bottles` and `num_exchange`, return _the **maximum** number of water bottles you can drink_.
+> <span style="float:right;">Source: [LeetCode 1518](https://leetcode.com/problems/water-bottles/description/)</span>
+
+---
+
+# Step 1: Understand the Problem
+
+- What variables are involved?
+    - $N_\mathrm{bottles}$: The number of _full_ water **bottles** that we currently have.
+    - $N_\mathrm{drunk}$: The number of _empty_ water bottles that we have **drunk**.
+- How are they related?
+    - The initial value for $N_\mathrm{bottles}$ is given by the input.
+    - When $N_\mathrm{bottles} \ge$ `num_exchange`, we can drink `num_exchange` of them, exchange them for one full water bottle, and then repeat the process.
+    - When $N_\mathrm{bottles} <$ `num_exchange`, we have to drink them up and quit.
+- What do we want to calculate?
+    - Given `num_bottles` and `num_exchange`, calculate the maximum number of water bottles we can drink.
+
+---
+
+# Step 2: Define the Interfaces
+
+> Given the two integers `num_bottles` and `num_exchange`, return _the **maximum** number of water bottles you can drink_.
+
+```moonbit
+fn num_water_bottles(num_bottles: Int, num_exchange: Int) -> Int {
+    abort("To be done")
+}
+```
+
+---
+
+# Step 3: Write the Test Cases
+
+```moonbit
+fn assert(assertion: Bool) {
+  if assertion.not() {
+    abort("Test failed")
+  }
+}
+
+test {
+  assert(num_water_bottles(9, 3) ==  13) // 9 + 3 + 1 = 13
+  assert(num_water_bottles(15, 4) ==  19)
+}
+```
+
+---
+
+# Step 4: Implement the Program
+
+A sample implementation:
+
+```moonbit
+fn num_water_bottles(num_bottles: Int, num_exchange: Int) -> Int {
+  fn consume(num_bottles, num_drunk) {
+    if num_bottles >= num_exchange {
+      let num_bottles -= num_exchange + 1
+      let num_drunk += num_exchange
+      consume(num_bottles, num_drunk)
+    } else {
+      num_bottles + num_drunk
+    }
+  }
+  consume(num_bottles, 0)
+}
+// Test cases omitted for brevity.
+```
+
+---
+
+# [Let's have a try!](https://try.moonbitlang.cn/#bc2238e7)
+
+---
+
+# Quiz
+
+- For some abnormal inputs, the sample program may fail. Can you identify them?
+    - Hint: In Java, the range of `int` values is $-2^{31}$ to $2^{31} - 1$.
+
+---
+
+# Summary
+
+- It is highly recommended to adopt a **test-driven development** (TDD) workflow, _viz._,
```

Comment:
I'm not sure if viz. is commmonly used enough or not. Cambridge dictionary marks it as 'old-fashioned'. So maybe just write 'that is'?

Reply:
I changed it to

```markdown
- It is recommended to adopt a **test-driven development** (TDD) workflow, namely,
```

This way we can still put the sentence in one line.